src/analysis.rs

```diff
@@ -0,0 +1,547 @@
+/// because the onepass DFA is implemented in the byte space, not the
+/// character space. This means, for example, that a branch between
+/// lowercase delta and uppercase delta is actually non-deterministic.
+fn first_byte(c: char) -> u8 {
+    let mut b: [u8; 4] = [0; 4];
+    c.encode_utf8(&mut b);
+    b[0]
+}
+
+/// A representation of all the possible ways a word in the language
+/// of a regex could begin. ClassBytes has no way to express the empty
+/// string, so we add an extra flag to indicate if a FirstSet includes
+/// epsilon. Put in a more theoretical way all firstsets are subsets of
+/// SIGMA `union` { epsilon }.
+#[derive(Debug, PartialEq, Eq)]
+struct FirstSet {
+    bytes: ClassBytes,
+    pub accepts_empty: bool,
+}
+
+impl FirstSet {
+    fn empty() -> Self {
+        FirstSet {
+            bytes: ClassBytes::empty(),
+            accepts_empty: false,
+        }
+    }
+
+    pub fn new<I>(ranges: I, accepts_empty: bool) -> Self
+    where I: IntoIterator<Item=ClassBytesRange>
+    {
+        FirstSet {
+            bytes: ClassBytes::new(ranges),
+            accepts_empty: accepts_empty,
+        }
+    }
+
+    fn anychar() -> FirstSet {
+        let mut f = FirstSet::empty();
+        f.push_bytes(ClassBytesRange::new(b'\0', b'\xFF'));
+        f
+    }
+
+    fn epsilon() -> FirstSet {
+        FirstSet {
+            bytes: ClassBytes::empty(),
+            accepts_empty: true,
+        }
+    }
+
+    fn push_bytes(&mut self, byte_range: ClassBytesRange) {
+        self.bytes.push(byte_range);
+    }
+
+    fn union(&mut self, other: &FirstSet) {
+        self.bytes.union(&other.bytes);
+        self.accepts_empty = self.accepts_empty || other.accepts_empty;
+    }
+
+    fn intersect(&mut self, other: &FirstSet) {
+        self.bytes.intersect(&other.bytes);
+        self.accepts_empty = self.accepts_empty && other.accepts_empty;
+    }
+
+    fn is_empty(&self) -> bool {
+        self.bytes.is_empty() && !self.accepts_empty
+    }
+
+    fn accepts_empty(&self) -> bool {
+        self.accepts_empty
+    }
+}
+
+/// An iterator over a concatenation of expressions which
+/// drills down into other embedded concatenations.
+struct NestedConcat<'a>(Vec<(&'a [Hir], usize)>);
+impl<'a> NestedConcat<'a> {
+    fn new(es: &'a [Hir]) -> Self {
+        NestedConcat(vec![(es, 0)])
+    }
+}
+impl<'a> Iterator for NestedConcat<'a> {
+    type Item = &'a Hir;
+
+    fn next(&mut self) -> Option<&'a Hir> {
+        if self.0.len() == 0 {
+            return None;
+        }
+
+        let tip = self.0.len() - 1;
+        let (es, idx) = self.0[tip];
+
+        if idx >= es.len() {
+            self.0.pop();
+            return self.next();
+        }
+
+        self.0[tip].1 += 1;
+
+        match es[idx].kind() {
+            &HirKind::Concat(ref es) => {
+                self.0.push((es, 0));
+                self.next()
+            }
+            _ => Some(&es[idx]),
+        }
```

Comment:
@ethanpailes I _believe_ this routine uses a stack size proportional to the depth of the HIR. It seems like it should be possible to flatten this with a loop.

Reply:
Nice catch. Thanks! I'll fix this with a loop as you suggest.

Note to future self: The degenerate case is when a bunch of concats are nested, with the child concat always coming first.